include/platform/mir/graphics/program_factory.h

```diff
@@ -41,16 +41,23 @@ class ProgramFactory
     /**
      * Compile and link a fragment-shader fragment into a full shader Program
      *
-     * \param extension_fragment [in]    Fragment to include necessary extensions
+     * \note    The ID is used to identify this shader program for caching purposes. Accidentally
+     *          using an ID which is already used by another shader will result in incorrect rendering.
+     *          We recommend using the address of a static local variable as the ID to guarantee
+     *          uniqueness.
+     *
+     * \param id [in]                   An opaque ID for this shader.
+     * \param extension_fragment [in]   Fragment to include necessary extensions
      * \param fragment_fragment [in]    The fragment-shader fragment. This must be a GLSL
-     *                                          fragment defining a function:
-     *                                          vec4 sample_to_rgba(in vec2 texcoord)
-     *                                          returning a vector containing the RGBA value at texcoord.
-     *                                          The elements of the uniform array tex[n] will be bound to
-     *                                          GL_TEXTURE0, GL_TEXTURE1, …, GL_TEXTURE(n-1). (n <= 8)
-     * \return  A fully compiled and linked Program
+     *                                  fragment defining a function:
+     *                                  vec4 sample_to_rgba(in vec2 texcoord)
+     *                                  returning a vector containing the RGBA value at texcoord.
+     *                                  The elements of the uniform array tex[n] will be bound to
+     *                                  GL_TEXTURE0, GL_TEXTURE1, …, GL_TEXTURE(n-1). (n <= 8)
+     * \return  A reference to the fully compiled and linked Program
      */
-    virtual std::unique_ptr<Program> compile_fragment_shader(
+    virtual Program& compile_fragment_shader(
+        void* id,
```

Comment:
@AlanGriffiths could this be the thing that causes the ABI break?

Reply:
That sure looks like an ABI break to me.